pkg/storage/redis.go

```diff
@@ -62,9 +68,56 @@ func (b *RedisDB) Close() error {
 	return b.db.Close()
 }
 
+func (b *RedisDB) Execute(ctx context.Context, businessLogicFunc BusinessLogicFunc, watchKeys []string) (any, error) {
+	var finalOutput any
+	// Transactional function.
+	txf := func(tx *goredislib.Tx) error {
+		// Operation is commited only if the watched keys remain unchanged.
+		_, err := tx.TxPipelined(ctx, func(pipe goredislib.Pipeliner) error {
+
+			pipeCTX := context.WithValue(ctx, TX, pipe)
```

Comment:
Passing parameters around with context is discouraged.  How about the following, so you don't need to create this additional context?

As an additional benefit, it becomes very clear to the users of the `storage.ServiceStorage` what functions they can use inside the transaction.

```go
// in storage.go

func BusinessLogicFunc func(ctx context.Context, tx Tx) (any, error)

type ServiceStorage interface {
  // ...
  Execute(ctx context.Context, businessLogicFunc BusinessLogicFunc, watchKeys []string) (any, error)
}

type Tx interface {
  Read(ctx, namespace, key string) ([]byte, error)
  Write(ctx context.Context, namespace, key string) error
}

// in redis.go
type redisTx struct {
  pipe goredislib.Pipeliner
}

func (tx *redisTx) Read(ctx, namespace, key string, watchKeys []string) ([]byte, error) {
  return tx.pipe.Get(ctx, getRedisKey(namespace, key))
}

func (tx *redisTx) Write(ctx context.Context, namespace, key string) error {
  return tx.pipe.Set(ctx, getRedisKey(namespace, key)).Err()
}

func (b *RedisDB) Exec(ctx context.Context, bFun BusinessLogicFunc) (any, error) {
  txf := func(tx *goredislib.Tx) error {
    _, err := tx.TxPipelined(ctx, func(pipe goredislib.Pipeliner) error {
      var err error
      finalOutput, err = bFun(ctx, redisTx{pipe})
      return err
    })
    return err
  }
  b.db.Watch(ctx, txf, watchKeys...)
}